### PR TITLE
Use optimized mesh scatter ops

### PR DESCRIPTION
**Pull request by peterdsharpe**
# PhysicsNeMo Pull Request

## Description

Refactors various Mesh ops to use optimized `scatter_aggregate` and `bincount()`. Creates a single target for performance improvements and simplifies codebase.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/mesh/geometry/dual_meshes.py [hunk 1]
```diff
@@ -54,40 +54,43 @@
 def _scatter_add_cell_contributions_to_vertices(
     dual_volumes: torch.Tensor,  # shape: (n_points,)
     cells: torch.Tensor,  # shape: (n_selected_cells, n_vertices_per_cell)
-    contributions: torch.Tensor,  # shape: (n_selected_cells,)
+    contributions: torch.Tensor,  # shape: (n_cells,) or (n_cells, n_vertices_per_cell)
```
Comment (laserkelvin) [inline physicsnemo/mesh/geometry/dual_meshes.py]: Why don't we use `jaxtyping` here? Mainly wanted to see if there was a technical blocking reason for this that I missed

I guess it's not done in this file, but would be good to track this in an issue if you want to leave it out of scope for this PR.

File: physicsnemo/mesh/geometry/dual_meshes.py [hunk 2]
```diff
@@ -54,40 +54,43 @@
...
-    n_vertices_per_cell = cells.shape[1]
-    for vertex_idx in range(n_vertices_per_cell):
-        dual_volumes.scatter_add_(
-            0,
-            cells[:, vertex_idx],
-            contributions,
-        )
+    if contributions.ndim == 1:
+        contributions = contributions.unsqueeze(-1).expand_as(cells)
+    dual_volumes.scatter_add_(0, cells.flatten(), contributions.reshape(-1))
```
Comment (laserkelvin) [inline physicsnemo/mesh/geometry/dual_meshes.py:99]: Since this is not public API maybe it's not so important, but just wondering if you think it would be good to just have a guard `if contributions.ndim not in (1, 2): raise ...`

**Comment by coreyjadams** [comment]
Any view into what kind of performance delta this gives @peterdsharpe ?

File: test/mesh/curvature/test_voronoi_tets.py [hunk 3]
```diff
@@ -29,6 +32,81 @@ def device():
...
+    def test_isolated_vertex(self, device):
+        """Vertex not in any edge gets zero dual volume."""
+        points = torch.tensor([[0.0], [1.0], [99.0]], device=device)
+        cells = torch.tensor([[0, 1]], dtype=torch.long, device=device)
+        mesh = Mesh(points=points, cells=cells)
+
+        dual_vols = compute_dual_volumes_0(mesh)
+        assert dual_vols[2] == 0.0
+
+
```
Comment (laserkelvin) [inline test/mesh/curvature/test_voronoi_tets.py:109]: Do we need to test against the 2D case as well?

**Review by laserkelvin — APPROVED** [review]
Generally LGTM, left some comments - I'll defer to your judgement :)

File: physicsnemo/mesh/geometry/dual_meshes.py (hunk 1 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/geometry/dual_meshes.py]: Very good call, will fix

File: test/mesh/curvature/test_voronoi_tets.py (hunk 3 above)
Comment (peterdsharpe) [reply test/mesh/curvature/test_voronoi_tets.py:109]: Good idea, will add

File: physicsnemo/mesh/geometry/dual_meshes.py (hunk 2 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/geometry/dual_meshes.py:99]: Doesn't hurt, I can add it

**Comment by peterdsharpe** [comment]
> Any view into what kind of performance delta this gives @peterdsharpe ?

So, this PR touches a few different items. Below is an auto-generated benchmark on AirFRANS meshes. 
That said, I think the bigger win is that as we add more performance code via `nn.functional`, we'll have fewer places that we need to update - so this will be a long-term maintainability benefit.

-------

### Performance benchmarks

Benchmarked on real AirFRANS 2D triangular CFD meshes (177K-365K cells single, up to 7M cells stacked) on CPU and NVIDIA RTX 4090 (CUDA), PyTorch 2.11. Median of 200 iterations, 10 warmup.

#### Change 1: Python loop over cell columns → single flattened `scatter_add_`

**Files:** `dual_meshes.py` (`_scatter_add_cell_contributions_to_vertices`, `compute_dual_volumes_0`)

This is the hot-path change - dual volumes feed into DEC operators and curvature computations.

| Mesh size | CPU (uniform/cell) | CPU (Meyer Voronoi) | GPU (uniform/cell) | GPU (Meyer Voronoi) |
|-----------|:--:|:--:|:--:|:--:|
| 177K cells | 1.12x | **2.11x** | 1.21x | **1.67x** |
| 344K cells | 1.38x | **2.39x** | 1.09x | **1.85x** |
| 1.76M cells | 1.59x | **2.16x** | 0.72x | **1.19x** |
| 7.05M cells | 1.00x | **2.12x** | 1.54x | **2.81x** |

The 2D per-vertex case (Meyer mixed Voronoi areas) is the main beneficiary: the old code ran a 3-iteration Python loop with 3 separate `scatter_add_` kernel launches; the new code does a single `scatter_add_` on flattened tensors.

#### Change 2: `zeros + scatter_add_(ones)` → `torch.bincount`

**Files:** `_topology.py`, `sample_data.py`, `laplacian.py`, `bvh.py`

| Mesh size | CPU | GPU |
|-----------|:--:|:--:|
| 177K cells | 1.25x | 0.38x |
| 344K cells | 1.15x | 0.40x |
| 1.76M cells | 1.85x | 0.81x |
| 7.05M cells | 1.65x | 0.83x |

`bincount` is faster on CPU (up to 1.85x) but slower on GPU (PyTorch's CUDA `bincount` uses a different kernel than `scatter_add_`). All call sites are cold paths: mesh topology validation (`is_manifold`, `is_watertight`), sharp-edge detection (once per `smooth_laplacian`), BVH candidate counting, and query containment counting. Absolute cost on a 180K-cell mesh is 0.08ms → 0.21ms per call.

#### Change 3: Manual `scatter_add_` + division → `scatter_aggregate` utility

**Files:** `_partition.py` (area sum, weighted-normal sum, weighted-centroid mean), `_scatter_ops.py`

| Mesh size | CPU (weighted mean) | GPU (weighted mean) | CPU (area sum) | GPU (area sum) |
|-----------|:--:|:--:|:--:|:--:|
| 177K cells | 1.06x | 0.95x | 1.25x | 0.88x |
| 344K cells | 1.05x | 0.96x | 1.08x | 0.84x |
| 1.76M cells | 0.85x | 0.99x | 0.99x | 0.99x |
| 7.05M cells | 1.00x | 1.02x | 1.01x | 1.10x |

Neutral performance. A fast path in `scatter_aggregate` for unweighted sums avoids unnecessary weight allocation.

#### Summary

The primary speedup comes from eliminating the Python-level loop in dual volume scatter (**1.2x-2.8x** on the hot path). The `bincount` and `scatter_aggregate` changes are code-quality improvements at negligible performance cost on cold paths.

**Review by coreyjadams — APPROVED** [review]
